File: mockit/src/main/scala/org/mockit/core/ConvertUtil.scala
package org.mockit.core

import scala.collection.mutable.ListBuffer

import org.mockit.core.BasicMockType.BasicMockType
import org.mockit.core.ConnectionType.ConnectionType
import org.mockit.General.{JFaultLevel, JConnectionType, JBasicMockType}
import org.mockit.mock.http.{HttpResponse, HttpRequest}

/**
 * Collection of implicit functions for java to scala conversion.
 *
 * @author  pheymann
 * @version 0.1.0
 */
object ConvertUtil {

    import org.mockit.networkclassloader.ClassConverter._
    
    implicit def javaToScala(jBasicType: JBasicMockType): BasicMockType = {
        jBasicType match {
            case JBasicMockType.client => BasicMockType.client
            case JBasicMockType.server => BasicMockType.server
            case JBasicMockType.agent  => BasicMockType.agent
            case JBasicMockType.p2p    => BasicMockType.p2p
            case JBasicMockType.none   => BasicMockType.none
        }
    }

    implicit def javaToScala(jConType: JConnectionType): ConnectionType = {
        jConType match {
            case JConnectionType.tcp    => ConnectionType.tcp
            case JConnectionType.udp    => ConnectionType.udp
            case JConnectionType.http   => ConnectionType.http
            case JConnectionType.none   => ConnectionType.none
        }
    }

    implicit def javaToScala(result: JMockResult): (Array[Byte], FaultLevel) = {
        (result.getResponse, javaToScala(result.getFault, result.getData))
    }

    implicit def javaToScala(config: JConfiguration): Configuration = {
        new Configuration(
            config.getThreadNumber,
            config.getRepetitions,
            config.getServerPort,
            config.getTargetPort,
            config.getTargetIp,
            DEFAULT_IP,
            config.getMockNumber,
            config.getMockType,
            config.getMockConnection
        )
    }

    implicit def javaToScala(jContainer: JMockUnitContainer): MockUnitContainer = {
        new MockUnitContainer(
            jContainer.getMockName,
            jContainer.getMockClass,
            jContainer.getConfig
        )
    }

    implicit def javaToScala(jMockUnits: java.util.List[JMockUnitContainer]): List[MockUnitContainer] = {
        val convert = new ListBuffer[MockUnitContainer]

        val iterator = jMockUnits.iterator

        while (iterator.hasNext) {
            convert += iterator.next
        }
        convert.toList
    }

    implicit def javaToScala(handler: java.util.function.Function[HttpRequest, HttpResponse])
                    : HttpRequest => HttpResponse = {
        request => handler.apply(request)
    }

    @throws(classOf[IllegalArgumentException])
    def javaToScala(jFault: JFaultLevel, jData: JFaultLevelData): FaultLevel = {
        jFault match {
            case JFaultLevel.noFault            => new NoFault
            case JFaultLevel.fixedDelay         => new FixedDelay(jData.getTime)
            case JFaultLevel.looseResponse      => new LooseResponse
            case JFaultLevel.looseConnection    => new LooseConnection
            case JFaultLevel.multipleResponses  => new MultipleResponses(jData.getFactor)
            case _ => throw new IllegalArgumentException("no level available")
        }
    }

}

File: mockit-java-api/src/main/java/com/github/pheymann/mockitjavaapi/core/ConvertUtil.java
package com.github.pheymann.mockitjavaapi.core;

import scala.Tuple2;
import scala.collection.mutable.ListBuffer;

import java.util.ArrayList;
import java.util.List;

import com.github.pheymann.mockitjavaapi.General;
import com.github.pheymann.mockit.core.BasicMockType;
import com.github.pheymann.mockit.core.ConnectionType;
import com.github.pheymann.mockit.core.FaultLevel;
import com.github.pheymann.mockit.core.*;
import com.github.pheymann.mockit.logging.LogEntry;

/**
 * Collection of functions for java to scala conversion.
 *
 * @author  pheymann
 * @version 0.1.0
 */
public class ConvertUtil {

    public static scala.Enumeration.Value javaToScala(final General.JBasicMockType jBasicType){
        if (jBasicType.equals(General.JBasicMockType.client))
            return BasicMockType.client();
        else if (jBasicType.equals(General.JBasicMockType.server))
            return BasicMockType.server();
        else if (jBasicType.equals(General.JBasicMockType.agent))
            return BasicMockType.agent();
        else if (jBasicType.equals(General.JBasicMockType.p2p))
            return BasicMockType.p2p();
        else if (jBasicType.equals(General.JBasicMockType.none))
            return BasicMockType.none();
        else
            return BasicMockType.none();
    }

    public static scala.Enumeration.Value javaToScala(final General.JConnectionType jConType) {
        if (jConType.equals(General.JConnectionType.tcp))
            return ConnectionType.tcp();
        else if (jConType.equals(General.JConnectionType.udp))
            return ConnectionType.udp();
        else if (jConType.equals(General.JConnectionType.http))
            return ConnectionType.http();
        else if (jConType.equals(General.JConnectionType.none))
            return ConnectionType.none();
        else
            return ConnectionType.none();
    }

    public static FaultLevel javaToScala(
        final General.JFaultLevel jFault,
        final JFaultLevelData jData
    ) throws IllegalArgumentException {
        if (jFault.equals(General.JFaultLevel.noFault))
            return new NoFault();
        else if (jFault.equals(General.JFaultLevel.fixedDelay))
            return new FixedDelay(jData.getTime());
        else if (jFault.equals(General.JFaultLevel.looseResponse))
            return new LooseResponse();
        else if (jFault.equals(General.JFaultLevel.looseConnection))
            return new LooseConnection();
        else if (jFault.equals(General.JFaultLevel.multipleResponses))
            return new MultipleResponses((jData.getFactor()));
        else
            throw new IllegalArgumentException("no valid fault level");
    }

    public static Tuple2<byte[], FaultLevel> javaToScala(JMockResult result) {
        return new Tuple2<>(result.getResponse(), javaToScala(result.getFault(), result.getData()));
    }

    public static List<LogEntry> scalaToJava(final ListBuffer<LogEntry> logs) {
        final scala.collection.Iterator<LogEntry> iterator = logs.iterator();

        final List<LogEntry> results = new ArrayList<>();

        while (iterator.hasNext()) {
            results.add(iterator.next());
        }
        return results;
    }

}
